### PR TITLE
Rework the banning and unbanning of entities in PolicyLists.

### PR DESCRIPTION
**Pull request by Gnuxie**
1. We keep track of the event that created a list rule so that we
can remove the rule by having a way to determine the original state key for the rule.
This is because the state key of rules can be anything and should not be
relied on by Mjolnir to unban things (which it was doing).

2. The old scheme for producing a state key was causing for some entities to escape bans
https://github.com/matrix-org/mjolnir/issues/322.

We could have used a hash or something similar, but we know that
the reason for the `rule:${entity}` scheme existed was for ease of debugging
and finding rules in devtools. So instead we have followed a scheme simalar to
bridges where the first character of an mxid is replaced with an underscore.
Everything else just gets put into the state key. Since domains can't have '@'
and room ids, aliases can't start with them either. (The reason you have to worry about
`@` is because state keys starting with `@` are reserved for users to put their own things there.)

3. We have stopped the need for Mjolnir to wait for the next response from sync after banning,
unbanning an entity so that we can apply ACL's sooner.
This also helps with attempts to turn Mjolnir into an appservice since appservices do not get informed of their own events via `/sync`

### CODE REVIEW

File: src/models/ListRule.ts [hunk 1]
```diff
@@ -78,6 +78,13 @@ const RECOMMENDATION_OPINION_VARIANTS: string[] = [
 export const OPINION_MIN = -100;
 export const OPINION_MAX = +100;
 
+interface MatrixStateEvent {
+    type: string,
+    content: any,
+    event_id: string,
+    state_key: string,
```
Comment (jesopo) [inline src/models/ListRule.ts:85]: i don't think i actually understand what state keys are

File: src/models/PolicyList.ts [hunk 2]
```diff
@@ -63,13 +63,27 @@ declare interface PolicyList {
 /**
  * The PolicyList caches all of the rules that are active in a policy room so Mjolnir can refer to when applying bans etc.
  * This cannot be used to update events in the modeled room, it is a readonly model of the policy room.
+ *
+ * You can inform the PolicyList about new events in a room with the `updateForEvent`, this will eventually
+ * cause the PolicyList to synchronize itself with the room if it doesn't know about that state event.
+ * Each time the PolicyList has finished synchronizing, it will emit the `'PolicyList.update'` event on itself as an EventEmitter.
```
Comment (jesopo) [inline src/models/PolicyList.ts:69]: what is it synchronising? its view of its own room?

File: src/models/PolicyList.ts [hunk 3]
```diff
@@ -63,13 +63,27 @@ declare interface PolicyList {
 /**
  * The PolicyList caches all of the rules that are active in a policy room so Mjolnir can refer to when applying bans etc.
  * This cannot be used to update events in the modeled room, it is a readonly model of the policy room.
+ *
+ * You can inform the PolicyList about new events in a room with the `updateForEvent`, this will eventually
```
Comment (jesopo) [inline src/models/PolicyList.ts:67]: there's a lot of overlapping terminology here; does this mean "you can inform a PolicyList object about new events in its respective policy room [...]"?

File: src/models/PolicyList.ts [hunk 4]
```diff
@@ -328,15 +373,26 @@ class PolicyList extends EventEmitter {
             }
         }
         this.emit('PolicyList.update', this, changes);
+        if (this.batchedEvents.keys.length !== 0) {
+            // The only reason why this isn't a TypeError is because we need to know about this when it happens, because it means
+            // we're probably doing something wrong, on the other hand, if someone messes with a server implementation and
+            // strange things happen where events appear in /sync sooner than they do in /state (which would be outrageous)
+            // we don't want Mjolnir to stop working properly. Though, I am not confident a burried warning is going to alert us.
```
Comment (jesopo) [inline src/models/PolicyList.ts:380]: this has come up before, and i think at some point we should think about external alerting for things that we really do need to see

**Review by jesopo — COMMENTED** [review]
looks great, just some questions about docs

**Comment by Gnuxie** [comment]
Test failures caused by #352

File: src/models/PolicyList.ts (hunk 2 above)
Comment (Gnuxie) [reply src/models/PolicyList.ts:69]: yeah i'll update this doc to say which method is doing that

File: src/models/PolicyList.ts (hunk 4 above)
Comment (Gnuxie) [reply src/models/PolicyList.ts:380]: Yepp!

File: src/models/PolicyList.ts (hunk 3 above)
Comment (Gnuxie) [reply src/models/PolicyList.ts]: i've tried to make things a bit clearer and explicit

**Review by jesopo — APPROVED** [review]

**Review by jesopo — APPROVED** [review]

**Pull request merged**